### ISSUE TITLE
Exception UX improvements...?

### ISSUE THREAD

**Issue by RubyNova** [novelrt/NovelRT#225]
As it stands, were currently using `std::runtime_error` for literally all our native exceptions, errors, etc.

Maybe we should consider Adding well-defined exceptions? Or at _least_ using some?

If we do decide to do this, it isn't urgent.

**Comment by Pheubel** [comment]
i'll have a look at some easy to pick out error

**Comment by RubyNova** [comment]
Kappa

**Comment by Pheubel** [comment]
Made a mock-up of how i would replace `std::runtime_error()` in AudioService,cpp

```c++
#ifndef NOVELRT_SERVICE_NOT_INITIALIZED_EXCEPTION_H
#define NOVELRT_SERVICE_NOT_INITIALIZED_EXCEPTION_H

#ifndef NOVELRT_H
#error Please do not include this directly. Use the centralised header (NovelRT.h) instead!
#endif

namespace NovelRt::Exceptions
{
  class ServiceNotInitializedException final : public std::runtime_error
  {
    public:
      ServiceNotInitializedException(const std::string functionName, const std::string message)
        : std::runtime_error("Unable to continue! Dangerous call being made to " + functionName + ". "+ message)
        {}
  };
} // namespace NovelRt::Exceptions

#endif //NOVELRT_SERVICE_NOT_INITIALIZED_EXCEPTION_H
```

The implementation would look something like this example:
```c++
void AudioService::pauseMusic() {
  if (!isInitialised) {
    _logger.logError("Cannot pause audio while the service is uninitialised! Aborting...");
    throw NovelRT::Exceptions::ServiceNotInitializedException("AudioService::pauseMusic()", "You cannot pause a sound when the AudioService is not initialised.");
  }

  alSourcePause(_musicSource);
}
```

Feel free to leave feedback, ideas and/or critique.

**Comment by RubyNova** [comment]
You have a typo in the namespace. I think you could also inherit from `std;;exception` instead?

**Comment by Pheubel** [comment]
> 
> 
> You have a typo in the namespace. I think you could also inherit from `std::exception` instead?

The reason i went with `std::runtime_error` is due to the different implementation of `std::exception` on MSVC, where there are both parameterless constructors as constructors with parameters. 

I made a small mock-up of a variant inheriting from `std::exception`
```c++
class ServiceNotInitializedException final : public std::exception
{
public:
	ServiceNotInitializedException(const std::string& function, const std::string& message)
		: std::exception("Unable to continue! Dangerous call being made to " + function + ". " + message)
	{}
};
```
this bit of code works just fine on my machine, but when i test it on [http://cpp.sh/67lal](http://cpp.sh/67lal) it can not find a matching function for std::exception(const char*)

`std::runtime_error` on the other hand is used the same way on different compilers, making it easier to make cross-platform custom exceptions.

references:
- [std::exception::exception](https://en.cppreference.com/w/cpp/error/exception/exception)
- [std::runtime_error](https://en.cppreference.com/w/cpp/error/runtime_error)

I might be missing something, so feel free to correct me.

ps: i also edited the previous mock-up

**Comment by RubyNova** [comment]
So I've had a deeper look - it seems `std::exception` just has more explicit handling for this via method overriding, so the question becomes "do we want this more explicit behaviour?"

**Comment by RubyNova** [comment]
Resolved by #241 .
